### PR TITLE
Process: Return created process object

### LINKED ISSUE #4104: Process: Change `Start Process` to return created process object instead of generic handle
Currently if you want to interact with the process started using the `Start Process` keyword you need to do it like this:
```
Start Process    command
${process} =    Get Process Object
# Do something with the process
```
The extra step needed to get the process object is somewhat annoying and it would be much more convenient to just use:
```
${process} =    Start Process    command
```
The problem with this is that `Start Process` currently returns a process handle that can be used with other keywords:
```
${handle} =    Start Process    command
# Other keywords possibly starting more processes
${result} =    Wait For Process    ${handle}
```
The handle is actually just an integer, and it ought to be possible to use the started process object as a handle instead. In practice the above would still work but `${handle}` would be a process object instead of an integer. This would nevertheless be a backwards incompatible change, because someone could have used integers as handles directly:
``` 
Start Process    command
# Other keywords possibly starting more processes
${result} =    Wait For Process    1
```
The above usage hasn't been recommended, though, and aliases should be used instead:
```
Start Process    command    alias=example
# Other keywords possibly starting more processes
${result} =    Wait For Process    example
```

### PR DESCRIPTION
**Pull request by onurcelep**
Modifies create_process method as it returns a process object.
Iterable for process objects  goes through connection indexes
so we can directly call get_connection for generated index from
register method.

Fixes #4104

### CODE REVIEW

**Comment by pekkaklarck** [comment]
Changing `Start Process` isn't enough, all keywords accepting `handle` need to be updated as well. For example,

```
${process} =    Start Process
${result} =    Wait For Process    ${process}
```
should work but this change alone breaks it. I'll add a comment how this can handled to the original issue.

This change would also need its own tests.

File: atest/testdata/standard_libraries/process/terminate_process.robot [hunk 1]
```diff
@@ -98,8 +98,9 @@ Terminate all processes
 Terminating all empties cache
     Some process
     Terminate All Processes    kill=True
-    ${handle} =    Some Process
-    Should Be Equal    ${handle}    ${1}
+    ${process_object}=    Some Process
+    ${index}=    Get Process Object Index     ${process_object}
+    Should Be Equal    ${index}    ${1}
```
Comment (pekkaklarck) [inline atest/testdata/standard_libraries/process/terminate_process.robot:103]: I don't think it's a good idea to add separate `Get Process Object Index` keyword for just this usage. Other alternatives to make sure the cache is empty:
1. Use `Switch Process    1` and handle resulting error by adding `[Documentation]` that specifies what error should occur. Although `Start Process` now returns Popen objects, `Switch Process` should work with indices. But not when there are no processes.
2. Use `BuiltIn.Get Library Instance` to get access to the library itself and check that `lib._processes` is empty. Accessing an attribute starting with `_` is a bit questionable but ought to be fine in tests.

File: src/robot/libraries/Process.py [hunk 2]
```diff
@@ -375,19 +375,32 @@ def start_process(self, command, *arguments, **configuration):
...
         for related examples.
 
-        Makes the started process new `active process`. Returns an identifier
-        that can be used as a handle to activate the started process if needed.
+        Makes the started process new `active process`. Returns the created 
+        process object [https://docs.python.org/3/library/subprocess.html#popen-constructor]
+        which can be assigned to a variable if needed.
+
+        Note: Returning an actual process object introduced in RF 5.0, previous versions
+        would only return a generic handler.
```
Comment (pekkaklarck) [inline src/robot/libraries/Process.py:383]: "handler" -> "process handle"

File: src/robot/libraries/Process.py [hunk 3]
```diff
@@ -375,19 +375,26 @@ def start_process(self, command, *arguments, **configuration):
...
 
         Processes are started so that they create a new process group. This
         allows sending signals to and terminating also possible child
         processes. This is not supported on Jython.
+
+        Examples:
+        | Start Process | ${command} | alias=example |
+        | ${result} = | Wait For Process | example |
+        | ${process} = | Start Process | python | -c | print('Hello, world!') |
+        | Should Be Equal | ${process.stdout} | Hello, world! |
```
[the remark this answers is not in the document]
Comment (pekkaklarck) [reply src/robot/libraries/Process.py:395]: Validation at the end should have `HELLO, WORLD!` instead of `Hello, world!`. Would be good to actually test the example as well!

File: src/robot/libraries/Process.py [hunk 4]
```diff
@@ -667,6 +680,19 @@ def get_process_object(self, handle=None):
...
+
+        If ``process`` is not given, uses the current `active process`.
+
+        Examples:
+        | Terminate All Processes |
+        | ${process_object}=      | Start Process | ${command} |
+        | ${process_index}=       | Get Process Object Index   | ${process_object} |
+        | Should Be Equal         | ${process_index}           | ${1}
+        """
+        return self._processes.get_connection_index(process)
```
Comment (pekkaklarck) [inline src/robot/libraries/Process.py:694]: I don't think anyone needs the old process handle (i.e. index) for anything. Apparently this keyword was added to make it fix a test that was broken when `Start Process` doesn't anymore return indices, but I already commented that there are other ways to validate that the internal cache is empty.

In other words, I believe this keyword should be removed. Can be added later if someone actually needs it.

File: src/robot/utils/connectioncache.py [hunk 5]
```diff
@@ -100,13 +100,28 @@ def get_connection(self, alias_or_index=None):
...
+        """
+        if popen is None:
+            if not self:
+                self.current.raise_error()
+            return self._connections.index(self.current)+1
+        try:
+            return self._connections.index(popen)+1
+        except ValueError:
+            raise ValueError(f"Provided input: '{popen}' is not a process object (popen).")
+
```
Comment (pekkaklarck) [inline src/robot/utils/connectioncache.py:124]: This code seems to be only used by the keyword that I believe isn't needed. Thus this code can be removed as well. If it is preserved, it should be made independent also on terminology level from Popen.

File: src/robot/utils/connectioncache.py [hunk 6]
```diff
@@ -137,13 +152,14 @@ def __len__(self):
     def __bool__(self):
         return self.current is not self._no_current
 
-    def resolve_alias_or_index(self, alias_or_index):
-        for resolver in self._resolve_alias, self._resolve_index:
+
+    def _resolve_connection(self, alias_or_index):
```
Comment (pekkaklarck) [inline src/robot/utils/connectioncache.py:156]: This method is part of the public API (see #3125) and thus cannot be renamed. Now that we support also connections themselves, `resolve_alias_or_index` is not a totally accurate name, but renaming it would require first a deprecation period. I personally consider this name good enough and would perhaps only update documentation. If you want to change the name, let's do that as a separate PR.

File: src/robot/utils/connectioncache.py [hunk 7]
```diff
@@ -137,13 +152,14 @@ def __len__(self):
...
+        for resolver in self._resolve_alias, self._resolve_index, self._is_connection:
             try:
                 return resolver(alias_or_index)
             except ValueError:
                 pass
-        raise ValueError("Non-existing index or alias '%s'." % alias_or_index)
+        raise ValueError(f"Non-existing index,alias or connection '{alias_or_index}'.")
```
Comment (pekkaklarck) [inline src/robot/utils/connectioncache.py:162]: Space needed after a comma (i.e. `index, alias`, not `index,alias`). I'd be actually fine considering connections as aliases which would mean we could preserve the old error message.

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
Looks good but some changes needed still. Please take a look at the individual comments for more information.

**Comment by pekkaklarck** [comment]
You mentioned that one acceptance test fails when you run tests locally and CI confirms that:
```
FAIL: Robot.Libdoc.Python Library.KwArgs and VarArgs
Process Should Be Stopped != Run Process
```
The reason is that this particular tests generates documentation for the Process library using Libdoc and validates that `Run Process` has expected argument information. Now that you have added a new keyword to the library, finding right keyword by an index fails. This could be fixed by incrementing the index, but I believe that this new keyword is unnecessary and it should be removed instead.

File: src/robot/utils/connectioncache.py (hunk 7 above)
Comment (onurcelep) [reply src/robot/utils/connectioncache.py]: reverted to old error msg

File: src/robot/utils/connectioncache.py (hunk 6 above)
Comment (onurcelep) [reply src/robot/utils/connectioncache.py]: reverted method name change , we can figure that in another PR

File: src/robot/utils/connectioncache.py (hunk 5 above)
Comment (onurcelep) [reply src/robot/utils/connectioncache.py]: removed index getter method

File: src/robot/libraries/Process.py (hunk 4 above)
Comment (onurcelep) [reply src/robot/libraries/Process.py]: got rid of `get_connection_index` method

File: src/robot/libraries/Process.py (hunk 2 above)
Comment (onurcelep) [reply src/robot/libraries/Process.py]: fixed typo

File: atest/testdata/standard_libraries/process/process_library.robot [hunk 8]
```diff
@@ -41,6 +41,12 @@ Input things to process
     ${result}=    Wait For Process
     Should Match    ${result.stdout}    *inp 42*
 
+Assign process object to variable
+    ${process} =  Start Process  python  -c  print('Hello, world!')
+    ${result} =  Run Process  python  -c  import sys; print(sys.stdin.read().upper().strip())  stdin=${process.stdout}
+    Wait For Process  ${process}
+    Should Be Equal As Strings  ${result.stdout}  HELLO, WORLD!
```
Comment (pekkaklarck) [inline atest/testdata/standard_libraries/process/process_library.robot:48]: Our convention is to use four spaces (or more if needed) between arguments and keywords.

File: atest/robot/standard_libraries/process/process_library.robot [hunk 9]
```diff
@@ -25,3 +25,6 @@ Input things to process
 
 Get process id
     Check Test Case    ${TESTNAME}
+
+Assign process object to variable
+    Check Test Case    ${TESTNAME}
```
Comment (pekkaklarck) [inline atest/robot/standard_libraries/process/process_library.robot:30]: Newline missing at the end of file. Good idea to configure your IDE to add them automatically.

This test should actually be before Get process id to keep them in same order as in the test data side.

File: atest/testdata/standard_libraries/process/terminate_process.robot [hunk 10]
```diff
@@ -96,10 +96,11 @@ Terminate all processes
     END
 
 Terminating all empties cache
+    [Documentation]    FAIL Non-existing index or alias '1'.
     Some process
+    Switch process  ${1}
     Terminate All Processes    kill=True
-    ${handle} =    Some Process
-    Should Be Equal    ${handle}    ${1}
+    Switch process  ${1}
```
Comment (pekkaklarck) [inline atest/testdata/standard_libraries/process/terminate_process.robot:103]: Good idea to validate that `Switch process    1` (`1` should work in addition to `${1}`) works before cache is cleared, but now it's possible that the test fails already there and test status validation doesn't detect that. To make sure it fails only at the end, we'd either need to validate that on the `atest/robot` side or change the test a bit. The latter is probably easier. Could, for example, start two processes and then use `Switch process    2` before clearing the cache.

**Review by pekkaklarck — APPROVED** [review]
Looks great! You still may want to take a look at my final and pretty nitpickey comments, but I consider this ready to be merged also without them.

File: atest/testdata/standard_libraries/process/terminate_process.robot (hunk 10 above)
Comment (onurcelep) [reply atest/testdata/standard_libraries/process/terminate_process.robot]: updated the test case as it creates 2 processes then clears cache

File: atest/robot/standard_libraries/process/process_library.robot (hunk 9 above)
Comment (onurcelep) [reply atest/robot/standard_libraries/process/process_library.robot]: moved the test case

File: atest/testdata/standard_libraries/process/process_library.robot (hunk 8 above)
Comment (onurcelep) [reply atest/testdata/standard_libraries/process/process_library.robot]: fixed

**Comment by pekkaklarck** [comment]
Looks great! Awesome how little changes were needed in the end.
